lib/kcl.js

```diff
@@ -153,41 +154,70 @@ module.exports = function(config, kinesis) {
-        if (err) throw err;
+    if (config.verbose) logger.info('[ getRecords ] params:', JSON.stringify({
+      ShardIterator: shard.iterator,
+      Limit: config.limit
+    }));
 
-        if (config.cloudwatchNamespace &&
-          resp.Records &&
-          resp.Records[0] &&
-          resp.Records[0].ApproximateArrivalTimestamp) {
-          throttledPutToCloudwatch('ShardIteratorAgeInMs', (+new Date()) - resp.Records[0].ApproximateArrivalTimestamp, 'Milliseconds', shard.id);
-        }
+    var req = kinesis.getRecords({ShardIterator: shard.iterator, Limit: config.limit});
+    var requestTimeOut = setTimeout(req.abort.bind(req), 5000);
 
-        // malformed responses will have an undefined NextShardIterator. In that case, we keep the existing one.
-        shard.iterator = resp.NextShardIterator === undefined ? shard.iterator : resp.NextShardIterator;
-        shard.lastGetRecords = +new Date();
+    req.on('error', function (err) {
+      if (err.code === 'SyntaxError' || err.code === 'RequestAbortedError') {
+        if (config.verbose) logger.info('[ getRecords ] Found SyntaxError', err);
+        return setTimeout(getRecords.bind(this, shard), 500);
+      }
+      if (err.code === 'ProvisionedThroughputExceededException') {
+        if (config.verbose) logger.info('[ getRecords ] Found ProvisionedThroughputExceededException', err);
+        return setTimeout(getRecords.bind(this, shard), 1000);
+      }
+      if (err.code === 'ServiceUnavailable' || err.code === 'InternalFailure' || err.code === 'NetworkingError') {
+        if (config.verbose) logger.info('[ getRecords ] Found serviceUnavailable', err);
+        return setTimeout(getRecords.bind(this, shard), 2000);
+      }
+      throw new Error(err);
 
-        if(resp.Records && resp.Records.length > 0) {
-          shard.sequenceNumber = resp.Records[resp.Records.length -1].SequenceNumber;
-          config.processRecords.call(shard, resp.Records, processRecordsDone);
+    });
+    req.on('extractData', function (response) {
+      if (config.verbose) logger.info('[ getRecords ] extractData', response.data.Records.length, 'items');
+    });
+    req.on('retry', function (resp) {
+      if (config.verbose) logger.info('[ getRecords retry ] retry count: ', resp.retryCount);
+    });
+    req.on('success', function (resp) {
+      if (config.verbose) logger.info('[ getRecords ] success Found', resp.data.Records.length, 'items');
+
+      if (config.cloudwatchNamespace &&
+        resp.Records &&
+        resp.Records[0] &&
+        resp.Records[0].ApproximateArrivalTimestamp) {
+        throttledPutToCloudwatch('ShardIteratorAgeInMs', (+new Date()) - resp.Records[0].ApproximateArrivalTimestamp, 'Milliseconds', shard.id);
+      }
+
+      // malformed responses will have an undefined NextShardIterator. In that case, we keep the existing one.
+      shard.iterator = resp.data.NextShardIterator === undefined ? shard.iterator : resp.data.NextShardIterator;
```

Comment:
I don't know if it's better to trigger a retry in that case.

This got rid of some errors but behaved weirdly in case of resharding (`resp.data.NextShardIterator` is [supposed to be](http://docs.aws.amazon.com/kinesis/latest/APIReference/API_GetRecords.html) `null` but the shard was not marked as complete).

This can be done separately from this PR though.